### PR TITLE
Finish converting 'Prepare' to MSBuild.

### PR DESCRIPTION
**Pull request by eerhardt**
Implement DownloadHostAndSharedFxArtifacts
Implement ZipTemplates
Switch the Prepare target to use all the MSBuild targets.

@brthor @sokket @piotrpMSFT @livarcocc 

Note: I copied a lot of stuff from #3696, will rebase/merge when that PR is in.

### CODE REVIEW

File: build/Microsoft.DotNet.Cli.Prepare.targets [hunk 1]
```diff
@@ -82,16 +85,18 @@
 
       <!-- Downloaded Installers + Archives -->
       <SharedHostCompressedFile>$(PackagesDirectory)/dotnet-host-$(ProductMonikerRid).$(SharedHostVersion)$(ArchiveExtension)</SharedHostCompressedFile>
```
Comment (brthor) [inline build/Microsoft.DotNet.Cli.Prepare.targets]: You mentioned removing the compressed files for shared host/hostfxr/sharedfx?

File: build/Microsoft.DotNet.Cli.Prepare.targets [hunk 2]
```diff
@@ -102,11 +107,80 @@
...
+
+  </Target>
+
+  <Target Name="ZipTemplates">
+    <PropertyGroup>
+      <_DotNetNewFolder>$(RepoRoot)/src/dotnet/commands/dotnet-new</_DotNetNewFolder>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <DotNetNewArchive Include="$([System.IO.Directory]::GetDirectories($(_DotNetNewFolder)))" />
```
Comment (brthor) [inline build/Microsoft.DotNet.Cli.Prepare.targets]: Does this search recursively or only top level?

Comment (eerhardt) [inline build/Microsoft.DotNet.Cli.Prepare.targets]: It does what it normall does in .NET:  https://msdn.microsoft.com/en-us/library/c1sez4sc(v=vs.110).aspx

File: build/Microsoft.DotNet.Cli.Prepare.targets (hunk 1 above)
Comment (eerhardt) [inline build/Microsoft.DotNet.Cli.Prepare.targets]: I'll do this in my next PR.

File: build/Microsoft.DotNet.Cli.Prepare.targets [hunk 3]
```diff
@@ -102,11 +107,80 @@
...
+    </ItemGroup>
+
+    <ArchiveDirectory InputDirectory="%(DotNetNewArchive.FullPath)"
+                      FileName="$([System.IO.Path]::GetFileName(%(DotNetNewArchive.FullPath)))"
+                      OutputDirectory="%(DotNetNewArchive.RelativeDir)"
```
Comment (brthor) [inline build/Microsoft.DotNet.Cli.Prepare.targets]: This is confusing me some, where does `RelativeDir` resolve to? Should there be some sort of `MakeRelative` Call here.

File: build_projects/dotnet-cli-build/build.ps1 [hunk 4]
```diff
@@ -71,7 +71,7 @@ $env:DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
 
 # Restore the build scripts
 Write-Host "Restoring Build Script projects..."
-pushd "$PSScriptRoot\.."
+pushd "$PSScriptRoot"
```
Comment (brthor) [inline build_projects/dotnet-cli-build/build.ps1]: doesn't this still need to restore it's p2p references?

File: build/Microsoft.DotNet.Cli.Prepare.targets (hunk 2 above)
Comment (brthor) [inline build/Microsoft.DotNet.Cli.Prepare.targets]: okay makes sense. I had only see `EnumerateDirectories` generally used.

File: build_projects/dotnet-cli-build/build.ps1 (hunk 4 above)
Comment (eerhardt) [inline build_projects/dotnet-cli-build/build.ps1]: There are none anymore. I changed it to all go in 1 assembly cuz MSBuild can't load dependencies strong next to this assembly.

Comment (brthor) [inline build_projects/dotnet-cli-build/build.ps1]: That's right, fair enough.

File: build/Microsoft.DotNet.Cli.Prepare.targets (hunk 3 above)
Comment (brthor) [inline build/Microsoft.DotNet.Cli.Prepare.targets]: Talked offline this should be doing the right thing.

**Comment by brthor** [comment]
:shipit:

File: Microsoft.DotNet.Cli.sln [hunk 5]
```diff
@@ -126,9 +126,20 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "dotnet-build3.Tests", "test
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "build", "build", "{89905EC4-BC0F-443B-8ADF-691321F10108}"
 	ProjectSection(SolutionItems) = preProject
+		build\Microsoft.DotNet.Cli.Package.targets = build\Microsoft.DotNet.Cli.Package.targets
 		build\Microsoft.DotNet.Cli.Prepare.targets = build\Microsoft.DotNet.Cli.Prepare.targets
 	EndProjectSection
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "package", "package", "{31A6D092-F50C-4D70-BA4D-97098189A953}"
+	ProjectSection(SolutionItems) = preProject
+		build\package\Microsoft.DotNet.Cli.Archive.targets = build\package\Microsoft.DotNet.Cli.Archive.targets
```
Comment (brthor) [inline Microsoft.DotNet.Cli.sln]: Do we get good VS editting on these? I've been on sublime

File: build/package/Microsoft.DotNet.Cli.Installer.MSI.targets [hunk 6]
```diff
@@ -2,7 +2,7 @@
 <Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
     
     <UsingTask TaskName="DownloadFile" AssemblyFile="$(CLIBuildDll)" />
-    <UsingTask TaskName="DecompressZip" AssemblyFile="$(CLIBuildDll)" />
+    <UsingTask TaskName="ExtractArchive" AssemblyFile="$(CLIBuildDll)" />
```
Comment (brthor) [inline build/package/Microsoft.DotNet.Cli.Installer.MSI.targets]: :+1: for name change

**Comment by brthor** [comment]
:shipit:

File: Microsoft.DotNet.Cli.sln (hunk 5 above)
Comment (eerhardt) [inline Microsoft.DotNet.Cli.sln]: I've been pretty good experience. Intellisense, squiggles on errors, etc

**Pull request merged**